ascent/examples/lists_using_recursive_enums.rs

```diff
@@ -0,0 +1,91 @@
+//! Lists using Recursive Enums
+//!
+//! The following example demonstrates the utilization of recursive records for
+//! building lists of strings over a given alphabet.
+//!
+//! The `List<T>` type is a recursive type where each instance is either `List::Nil`
+//! or a pair of a leading `T` and a tailing `List::Cons`.
+//!
+//! The relation `list` is defined to contain all lists of length 5 or less
+//! over a given alphabet defined by the relation `char`.
+//! The relation `len` is essentially a function assigning each list its length.
+//!
+//! Finally, the `res` relation illustrates how to create constant values for recursive record types.
+//!
+//! (Adapted from https://souffle-lang.github.io/examples#sequences-using-recursive-records)
+
+use std::rc::Rc;
+
+use ascent::ascent;
+
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub enum List<T> {
+  Cons(T, Rc<List<T>>),
+  Nil
+}
+
+macro_rules! cons {
+    ($h: expr, $t: expr) => {
+        Rc::new(List::Cons($h, $t))
+    }
+}
+
+macro_rules! nil {
+    () => {
+        Rc::new(List::Nil)
+    }
+}
+
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub struct Res(&'static str);
+
+ascent! {
+    // Facts:
+
+    relation char(char);
+    
+    // Rules:
+    
+    relation list(Rc<List<char>>);
+    
+    list(nil!());
+    list(cons!(c.clone(), l.clone())) <-- char(c), list(l), len(l, n), if *n < 5;
+
+    relation len(Rc<List<char>>, usize);
+    
+    len(nil!(), 0);
+    len(Rc::clone(&l), n + 1) <-- char(c), len(r, n), let l = cons!(c.clone(), r.clone()), list(Rc::clone(&l));
```

Comment:
Not quite, unfortunately:

```rust
error[E0507]: cannot move out of `l`, a captured variable in an `FnMut` closure
  --> ascent/examples/lists_using_recursive_enums.rs:58:9
   |
58 | ...en(l, n + 1) <-- char(c), len(r, n), let l = cons!(c.clone(), r.clone()), list(&...
   |       ^                                     -                                ---- captured by this `FnMut` closure
   |       |                                     |
   |       |                                     captured outer variable
   |       move occurs because `l` has type `Rc<List<char>>`, which does not implement the `Copy` trait
```

Reply:
hmm, ok. I'll have to look into it. Meanwhile, this should work:
```
len(l.clone(), n + 1) <-- char(c), len(r, n), let l = cons!(c.clone(), r.clone()), list(&l);
```

Reply:
Works! Fixed.

Reply:
Ok, I see why that wouldn't work. The last clause could have multiple matches in general (not the case here), so we can't take ownership of the value in the head clause.

Reply:
Yeah, that makes a lot of sense.